command.go

```diff
@@ -317,7 +317,29 @@ func (c *Command) parseFlags(args Args, shellComplete bool) (*flag.FlagSet, erro
 		return set, set.Parse(append([]string{"--"}, args.Tail()...))
 	}
 
-	err = parseIter(set, c, args.Tail(), shellComplete)
+	for pCtx := ctx.parentContext; pCtx != nil; pCtx = pCtx.parentContext {
+		if pCtx.Command != nil {
+			for _, fl := range pCtx.Command.Flags {
+				if pfl, ok := fl.(PersistentFlag); ok && pfl.IsPersistent() {
+					applyPersistentFlag := true
+					set.VisitAll(func(f *flag.Flag) {
+						for _, name := range fl.Names() {
+							if name == f.Name {
+								applyPersistentFlag = false
+							}
+						}
+					})
+					if applyPersistentFlag {
+						if err := fl.Apply(set); err != nil {
+							return nil, err
+						}
+					}
+				}
+			}
+		}
+	}
```

Comment:
Inverting some conditions and breaking early to reduce nesting

```suggestion
	for pCtx := ctx.parentContext; pCtx != nil; pCtx = pCtx.parentContext {
		if pCtx.Command == nil {
			continue
		}

		for _, fl := range pCtx.Command.Flags {
			pfl, ok := fl.(PersistentFlag)
			if !ok || !pfl.IsPersistent() {
				continue
			}

			applyPersistentFlag := true
			set.VisitAll(func(f *flag.Flag) {
				for _, name := range fl.Names() {
					if name == f.Name {
						applyPersistentFlag = false
					}
				}
			})

			if !applyPersistentFlag {
				continue
			}

			if err := fl.Apply(set); err != nil {
				return nil, err
			}
		}
	}
```

Reply:
Yes good idea.